src/esp/gfx/DebugLineRender.cpp

```diff
@@ -0,0 +1,314 @@
+  if (dist0 < radius) {
+    if (dist1 < radius) {
+      return false;
+    }
+    const float lerpFraction = (radius - dist0) / (dist1 - dist0);
+    CORRADE_INTERNAL_ASSERT(lerpFraction >= 0.f && lerpFraction <= 1.f);
+    Mn::Vector3 clippedPt = Mn::Math::lerp(*pt0, *pt1, lerpFraction);
+    *pt0 = clippedPt;
+    return true;
+  }
+
+  if (dist1 < radius) {
+    const float lerpFraction = (radius - dist1) / (dist0 - dist1);
+    CORRADE_INTERNAL_ASSERT(lerpFraction >= 0.f && lerpFraction <= 1.f);
+    Mn::Vector3 clippedPt = Mn::Math::lerp(*pt1, *pt0, lerpFraction);
+    *pt1 = clippedPt;
+    return true;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+DebugLineRender::DebugLineRender()
+    : _glResources{std::make_unique<GLResourceSet>()} {
+  _glResources->mesh.addVertexBuffer(_glResources->buffer, 0,
+                                     Mn::Shaders::VertexColor3D::Position{},
+                                     Mn::Shaders::VertexColor3D::Color4{});
+}
+
+void DebugLineRender::releaseGLResources() {
+  _glResources = nullptr;
+}
+
+void DebugLineRender::drawLine(const Mn::Vector3& from,
+                               const Mn::Vector3& to,
+                               const Mn::Color4& fromColor,
+                               const Mn::Color4& toColor) {
+  VertexRecord v1{from, remapAlpha(fromColor)};
+  VertexRecord v2{to, remapAlpha(toColor)};
+  arrayAppend(_verts, {v1, v2});
+}
+
+void DebugLineRender::setLineWidth(float lineWidth) {
+  // This is derived from experiments with glLineWidth on Nvidia hardware.
+  const float maxLineWidth = 20.f;
+  if (lineWidth > maxLineWidth) {
+    LOG(WARNING) << "DebugLineRender::setLineWidth: requested lineWidth of "
+                 << lineWidth
+                 << " is greater than "
+                    "max supported width of "
+                 << maxLineWidth;
+    lineWidth = maxLineWidth;
+  }
+  _internalLineWidth = lineWidth / 2;  // see also DebugLineRender::flushLines
+}
+
+void DebugLineRender::flushLines(const Magnum::Matrix4& camMatrix,
+                                 const Magnum::Matrix4& projMatrix,
+                                 const Magnum::Vector2i& viewport) {
+  CORRADE_ASSERT(_glResources,
+                 "DebugLineRender::flushLines: no GL resources; see "
+                 "also releaseGLResources", );
+
+  bool doToggleBlend = !glIsEnabled(GL_BLEND);
+
+  if (doToggleBlend) {
+    Mn::GL::Renderer::enable(Mn::GL::Renderer::Feature::Blending);
+    Mn::GL::Renderer::setBlendFunction(
+        Mn::GL::Renderer::BlendFunction::SourceAlpha,
+        Mn::GL::Renderer::BlendFunction::OneMinusSourceAlpha);
+    Mn::GL::Renderer::setBlendEquation(Mn::GL::Renderer::BlendEquation::Add);
+  }
+
+  // Update buffer with new data
+  _glResources->buffer.setData(_verts, Mn::GL::BufferUsage::DynamicDraw);
+
+  // Update shader
+  _glResources->mesh.setCount(_verts.size());
+
+  glDisable(GL_LINE_SMOOTH);  // anti-aliased lines cause artifacts
```

Comment:
Feels weird calling OpenGL directly here.

Reply:
I thought this was some ancient relic and so this isn't even exposed anywhere.

If you want a "less" direct way and something that should be more futureproof, `GL::Renderer::disable(GL::Renderer::Feature(GL_LINE_SMOOTH))` and then please add a TODO for me.

Reply:
I also hit an issue with this not even compiling for Emscripten. I'm starting to think it's pretty unlikely that anyone is going to set this setting to true. I'm just going to remove this line and not worry about it.

Reply:
Ah, *that's* why I thought it's an ancient relic, because GLES/WebGL doesn't even define such a thing :D